OSDOCS-1997: About Butane docs

bobfuru opened the pull request:
[OSDOCS-1997](https://issues.redhat.com/browse/OSDOCS-1997) - Adds a new module for an overview of Butane, including download instructions from mirror.openshift.com and a basic YAML example based on information from coreos/butane (https://coreos.github.io/butane/getting-started/). 

**Preview link:**
_Configuring machine configs with Butane_
https://deploy-preview-33794--osdocs.netlify.app/openshift-enterprise/latest/installing/install_config/installing-customizing.html#installation-special-config-butane_installing-customizing

bobfuru commented on modules/installation-special-config-butane.adoc:
```diff
@@ -0,0 +1,105 @@
+$ curl https://mirror.openshift.com/pub/openshift-v4/clients/butane/latest/butane-aarch64
+----
++
+. Unpack and unzip the archive.
```
It's not clear to me what to do from this point. I'm using steps from [Installing the OpenShift CLI](https://docs.openshift.com/container-platform/4.7/cli_reference/openshift_cli/getting-started-cli.html) for a workflow because I couldn't find an example for using the mirror site. Not sure what format the butane file is in. I got it downloaded using podman pull but my understanding is that we aren't using that in 4.8. 

Need some clarification of this process, please.

bobfuru commented on modules/installation-special-config-butane.adoc:
```diff
@@ -0,0 +1,105 @@
+
+.Prerequisites
+
+* You have added your local SSH private key to the `ssh-agent`. See _Generating an SSH private key and adding it to the agent_ in the "Installing" section.
```
Is this necessary or helpful? I can't xref specifically to the procedure cuz it's a module, but the workflow I'm using partly comes from here: https://docs.openshift.com/container-platform/4.7/installing/installing_bare_metal/installing-bare-metal.html#ssh-agent-using_installing-bare-metal

bobfuru commented on modules/installation-special-config-butane.adoc:
```diff
@@ -0,0 +1,105 @@
+* You have added your local SSH private key to the `ssh-agent`. See _Generating an SSH private key and adding it to the agent_ in the "Installing" section.
+* You have installed the `butane` utility.
+
+.Procedure
```
Another question is whether we actually want or need to include this basic config example? We have use case-specific examples elsewhere in the Install Config doc in the parent assembly.

bgilbert replied:
I'm not sure; maybe not.  If we do, maybe the example should include a non-Ignition field like the MachineConfig `kernelType`?

```yaml
openshift:
  kernel_type: realtime
```

We could link to the [full spec](https://coreos.github.io/butane/config-openshift-v4_8/) but that'd be an off-site link.

bobfuru replied:
Added `storage.filesystems` but feel free to suggest revisions for clarity or accuracy.

bobfuru commented on modules/installation-special-config-butane.adoc:
```diff
@@ -0,0 +1,105 @@
+
+. Create a Butane config file, such as `99-simple-example.bu`, as shown in the following example:
++
+[source,yaml]
```
The YAML is based on https://coreos.github.io/butane/getting-started/

bobfuru commented on modules/installation-special-config-butane.adoc:
```diff
@@ -0,0 +1,105 @@
+{"ignition":{"config":{"replace":{"source":null,"verification":{}}},"security":{"tls":{}},"timeouts":{},"version":"3.0.0"},"passwd":{"users":[{"name":"core","sshAuthorizedKeys":["ssh-rsa ssh-rsa AAAAB3NzaC1yc..."]}]},"storage":{},"systemd":{}}
+----
++
+An Ignition config is created. You can now provide the Ignition config to a booting {op-system} machine for Ignition to finish configuring your machines.
```
If we include this example, should we add the step to generate the YAML file (`butane 99-simple-example.bu --files-dir . -o 99-simple-example.yaml`)?

bgilbert replied:
With the `openshift` `4.8.0` spec, Butane generates MachineConfigs directly, and we should update the doc throughout to reflect this.

`--files-dir` is only relevant if `storage.trees` or `storage.files[].contents.local` is specified in the config, and should be omitted otherwise for clarity.

bobfuru commented on modules/rhcos-about.adoc:
```diff
@@ -89,7 +89,7 @@ The Ignition facility runs only when the {op-system} system is first set up. Aft
 
 Ignition is the utility that is used by {op-system} to manipulate disks during initial configuration. It completes common disk tasks, including partitioning disks, formatting partitions, writing files, and configuring users. On first boot, Ignition reads its configuration from the installation media or the location that you specify and applies the configuration to the machines.
 
```
Changes from line 91..* are wording revisions and do not need technical review.

bobfuru commented:
Hi @bgilbert - Here's a shot at the Butane overview and install doc. I left a # of comments/questions because I'm not clear about the install process using mirror.openshift.com and a few other areas. PTAL. Thanks!

bobfuru commented:
/hold

bgilbert commented on modules/installation-special-config-butane.adoc:
```diff
@@ -0,0 +1,105 @@
+variant: openshift
+version: 4.8.0
+metadata:
+  name: 99-simple-example
```
This example won't transpile because it's missing a mandatory metadata field:

```yaml
metadata:
  name: 99-simple-example
  labels:
    machineconfiguration.openshift.io/role: worker
```

bgilbert commented on modules/installation-special-config-butane.adoc:
```diff
@@ -0,0 +1,105 @@
+----
+<1> Set the `ssh_authorized_keys` value to be your SSH public key, which is usually the contents of the `~/.ssh/id_rsa.pub` file on your system.
+
+. Set your ssh public key. On Linux systems, this is usually the contents of the `~/.ssh/id_rsa.pub` file.
```
This doesn't seem relevant.

bgilbert reviewed (commented):
Thanks for handling this!  Should we also link back to this section from the other sections that use Butane?

bobfuru commented on installing/install_config/installing-customizing.adoc:
```diff
@@ -28,6 +28,7 @@ ifdef::openshift-webscale[]
 include::modules/installation-special-config-rtkernel.adoc[leveloffset=+1]
 endif::openshift-webscale[]
 include::modules/installation-special-config-kmod.adoc[leveloffset=+1]
+include::modules/installation-special-config-butane.adoc[leveloffset=+1]
```
Yes, moving it to the top makes more sense. I didn't like having it come after kmod, as you pointed out.

bgilbert commented on modules/installation-special-config-butane.adoc:
```diff
@@ -0,0 +1,116 @@
+
+== About Butane
+
+Butane is a YAML utility that {product-title} uses to provide convenient short-hand syntax for writing machine-readable machine configs, as well as for performing additional validation of  machine configs. The format of the Butane config file that Butane transpiles is defined in the
```
Some small wording improvements:

```suggestion
Butane is a command-line utility that {product-title} uses to provide convenient short-hand syntax for writing machine configs, as well as for performing additional validation of machine configs. The format of the Butane config file that Butane accepts is defined in the
```

bgilbert commented on modules/installation-special-config-butane.adoc:
```diff
@@ -0,0 +1,116 @@
+[id="rhcos-installing-butane_{context}"]
+== Installing Butane
+
+You can install the Butane tool (`butane`) to create {product-title} machine configs from a command-line interface. You can install `butane` on Linux, Windows, or macOS by downloading the related binary file.
```
Maybe s/related/corresponding/?

bgilbert commented on modules/installation-special-config-butane.adoc:
```diff
@@ -0,0 +1,116 @@
+$ curl https://mirror.openshift.com/pub/openshift-v4/clients/butane/latest/butane --output butane
+----
++
+.. Optional: For a specific type of architecture you are installing Butane on, such as aarch64 or ppc641e, indicate the appropriate URL. For example:
```
```suggestion
.. Optional: For a specific type of architecture you are installing Butane on, such as aarch64 or ppc64le, indicate the appropriate URL. For example:
```

bgilbert commented on modules/installation-special-config-butane.adoc:
```diff
@@ -0,0 +1,116 @@
+[id="rhcos-creating-with-butane_{context}"]
+== Creating a `MachineConfig` object by using Butane
+
+By using Butane, you can produce a `MachineConfig` object so that you can configure worker or master nodes at installation time or via the Machine Config Operator.
```
s/master/control plane/?

bgilbert commented on modules/installation-special-config-butane.adoc:
```diff
@@ -0,0 +1,116 @@
+
+. Create a Butane config YAML file. The following example tells the {product-title} cluster to use a real-time kernel for all worker nodes:
++
+[source,terminal]
```
For improved syntax highlighting, I've been putting the filename (`99-worker-realtime.bu`) in the bullet text above, then using `[source,yaml]` here and dropping the `cat`/`EOF` lines.

bobfuru replied:
Cool, I originally had `[source,yaml]` but talked myself into combining steps. I've edited to more closely match your `99-simple-kmod.bu` example.

bgilbert commented on modules/installation-special-config-butane.adoc:
```diff
@@ -0,0 +1,116 @@
+  labels:
+    machineconfiguration.openshift.io/role: worker
+openshift:
+    kernel_type: realtime
```
Two-space indent for consistency.

bgilbert commented on modules/installation-special-config-butane.adoc:
```diff
@@ -0,0 +1,116 @@
+----
++
+A `MachineConfig` object YAML file is created for you to finish configuring your machines.
+. Save the Butane config in case you need to update the manifest in the future.
```
s/manifest/`MachineConfig` object/ for consistency.

bgilbert commented on modules/installation-special-config-butane.adoc:
```diff
@@ -0,0 +1,116 @@
++
+[source,terminal]
+----
+$ butane $HOME/<cluster_config_dir>/99-worker-realtime.bu -o ./99-worker-realtime.yaml
```
We didn't tell them where to put the Butane config, so maybe we should omit the directory part of the input argument?

bobfuru replied:
Yes, agreed

bgilbert commented on installing/install_config/installing-customizing.adoc:
```diff
@@ -22,13 +22,19 @@ when installing bare-metal nodes.
 endif::openshift-webscale[]
 include::modules/installation-special-config-kmod.adoc[leveloffset=+1]
+
+See xref:../installing-customizing.adoc#installation-special-config-butane_installing-customizing[Configuring machine configs with Butane] for information on Butane.
```
Should these lines be moved inside the modules, before we tell the user to run Butane, rather than afterward?  That might be better even if we can't cross-link.

bobfuru reviewed (commented):
Thanks for the feedback, @bgilbert. 😀  I've reworked the references as you suggested and I believe I've resolved your other suggestions. One question inline about `realtime` vs. `default` in the note on line 97 of `modules/installation-special-config-butane.adoc`.

bobfuru commented:
@miabbott PTAL for QE review. 🙇

miabbott commented on modules/installation-special-config-butane.adoc:
```diff
@@ -0,0 +1,118 @@
+[id="installation-special-config-butane_{context}"]
+= Creating machine configs with Butane
+
+Machine configs are used to deploy control plane and worker machines by telling machines how to create users and filesystems, set up the network, install systemd units, and more.
```
I think using "deploy" here is the wrong verb.  When I think of deploying machines, I think of either installing the cluster from scratch, scaling up MachineSets, or manually adding nodes to the cluster.

I would prefer to say that machine configs are used to *configure* machines.

bobfuru replied:
Good clarification...resolved.

adellape reviewed (commented):
General comment: consider breaking up the new `modules/installation-special-config-butane.adoc` file into a concept and multiple procedure modules.

bgilbert commented on modules/installation-special-config-butane.adoc:
```diff
@@ -0,0 +1,118 @@
+$ echo $PATH
+----
++
+. Make the downloaded binary file executable:
```
Let's move this before the previous step, since `chmod +x butane` assumes the binary is still in the current directory and we've just moved it out.

bobfuru replied:
Good call!

bgilbert commented on modules/installation-special-config-butane.adoc:
```diff
@@ -0,0 +1,118 @@
++
+[source,terminal]
+----
+$ butane <butane_file.bu>
```
Yeah, I agree with dropping the `.bu` here.

bobfuru commented on modules/installation-special-config-kmod.adoc:
```diff
@@ -377,7 +377,12 @@ $ make install DESTDIR=${FAKEROOT}/usr/local CONFDIR=${FAKEROOT}/etc/
++
+[NOTE]
+====
+See _Creating machine configs with Butane_ for information about Butane.
```
Always gets me because I see our docs using both styles. Will default to quotes now. Thanks.

bobfuru commented on modules/installation-special-config-butane.adoc:
```diff
@@ -0,0 +1,118 @@
++
+[source,yaml]
+----
+variant: openshift
+version: 4.8.0
+metadata:
+  name: 99-worker-realtime
+  labels:
+    machineconfiguration.openshift.io/role: worker
+openshift:
+  kernel_type: realtime
+storage
+  filesystems:
+    - device: /dev/mapper/root
+      format: ext4
```
Thank you! Much classier than anything I could come up with. :)

bobfuru replied:
Also, I changed the name and repeated occurrences throughout to `99-worker-console` (s/realtime/console).

bobfuru commented on modules/rhcos-about.adoc:
```diff
@@ -133,9 +129,9 @@ The Ignition process for an {op-system} machine in an {product-title} cluster in
 * Ignition runs the Ignition config files to set up users, systemd unit files, and other configuration files.
 * Ignition unmounts all components in the permanent system that were mounted in the initramfs.
-* Ignition starts up new machine’s init process which, in turn, starts up all other services on the machine that run during system boot.
+* Ignition starts up the init process of the new machine, which in turn starts up all other services on the machine that run during system boot.
```
Thanks. I'm inclined to leave it for that reason but will be considering some improvements post-GA and will look at updating to reflect ISG.

bobfuru commented on modules/installation-special-config-butane.adoc:
```diff
@@ -0,0 +1,118 @@
+$ oc create -f 99-worker-realtime.yaml
+----
+
+For more detailed Butane config examples, see _Adding kernel modules to nodes_ and _Encrypting or mirroring disks during installation_.
```
Yes, that makes more sense!

bobfuru commented on modules/installation-special-config-butane.adoc:
```diff
@@ -0,0 +1,118 @@
++
+A `MachineConfig` object YAML file is created for you to finish configuring your machines.
+. Save the Butane config in case you need to update the `MachineConfig` object in the future.
+. If the cluster is not up yet, generate manifest files and add the `MachineConfig` object YAML file to the `openshift` directory. If the cluster is already running, apply the file as follows:
```
running++

bobfuru commented on modules/installation-special-config-butane.adoc:
```diff
@@ -0,0 +1,118 @@
+[id="installation-special-config-butane_{context}"]
+= Creating machine configs with Butane
+
+Machine configs are used to configure control plane and worker machines by telling machines how to create users and filesystems, set up the network, install systemd units, and more.
```
I never which way to land with backticks for systemd, etcd, initramfs, etc. There are several occurrences without the markup, so I'm going to leave it bare for now. But def a good topic for the next Great Style Debate.

bobfuru commented on modules/rhcos-about.adoc:
```diff
@@ -89,7 +89,7 @@ The Ignition facility runs only when the {op-system} system is first set up. Aft
 Ignition is the utility that is used by {op-system} to manipulate disks during initial configuration. It completes common disk tasks, including partitioning disks, formatting partitions, writing files, and configuring users. On first boot, Ignition reads its configuration from the installation media or the location that you specify and applies the configuration to the machines.
 
-Whether you are installing your cluster or adding machines to it, Ignition always performs the initial configuration of the {product-title} cluster machines. Most of the actual system setup happens on each machine itself. For each machine, Ignition takes the {op-system} image and boots the {op-system} kernel. Options on the kernel command line, identify the type of deployment and the location of  ...
+Whether you are installing your cluster or adding machines to it, Ignition always performs the initial configuration of the {product-title} cluster machines. Most of the actual system setup happens on each machine itself. For each machine, Ignition takes the {op-system} image and boots the {op-system} kernel. Options on the kernel command line identify the type of deployment and the location of t ...
```
See my next comment...going to leave it as is for now even though I don't disagree

bobfuru commented:
Thanks for the detailed reviews, @bgilbert and @adellape! I believe I've addressed all the feedback except for the mod docs modularization. My plan is to break the files into smaller modules (1 concept, 2 procedures) after SME/QE approval if there's time ahead of doc freeze.

@bgilbert and @miabbott Could you PTAL for LGTM or additional changes by EOD Friday? Many thanks!!!

bobfuru commented on modules/installation-special-config-butane.adoc:
```diff
@@ -0,0 +1,128 @@
+== About Butane
+
+Butane is a command-line utility that {product-title} uses to provide convenient, short-hand syntax for writing machine configs, as well as for performing additional validation of machine configs. The format of the Butane config file that Butane accepts is defined in the
+https://coreos.github.io/butane/specs/[OpenShift Butane config spec].
```
Rather than specify the OpenShift 4.8.0 spec, I went up a dir-level to keep it green.

bgilbert commented on modules/installation-special-config-kmod.adoc:
```diff
@@ -278,7 +278,7 @@ when you build your Ignition config.
 
 [id="provision-kernel-modules-via-machineconfig_{context}"]
-=== Provision kernel modules via a `MachineConfig` object
+=== Provision kernel modules via a `MachineConfig object
```
Missing close quote.

bobfuru replied:
Oops, thanks! (I removed the open quote as well here as part of our docs guidelines--[no markup for headings](https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#:~:text=as%20a%20heading.-,Do%20not%20use%20backticks%20or%20other%20markup%20in%20assembly%20or%20module%20headings.,-Discrete%20headings).)

bgilbert reviewed (approved):
LGTM!

bobfuru commented:
> General comment: consider breaking up the new `modules/installation-special-config-butane.adoc` file into a concept and multiple procedure modules.

Thanks, @adellape - I have broken this up into an h1 concept (main), h2 concept (about), and 2 procedures.

bobfuru commented:
/cherrypick enterprise-4.8
